File: Lexer.cpp
#include "Lexer.h"
#include <cctype>

Lexer::Lexer(const string& source) : source(source), currentIndex(0), currentChar(source[0]) {}

vector<Token> Lexer::tokenize() {
    vector<Token> tokens;
    while (!isAtEnd()) {
        skipWhitespace();
        if (isalpha(currentChar)) {
            tokens.push_back(lexIdentifier());
        } else if (isdigit(currentChar)) {
            tokens.push_back(lexNumber());
        } else {
            tokens.push_back(lexOperator());
        }
    }
    return tokens;
}

char Lexer::advance() {
    currentChar = source[++currentIndex];
    return currentChar;
}

char Lexer::peek() {
    if (isAtEnd()) return '\0';
    return source[currentIndex + 1];
}

bool Lexer::isAtEnd() {
    return currentIndex >= source.length();
}

void Lexer::skipWhitespace() {
    while (isspace(currentChar) && !isAtEnd()) {
        advance();
    }
}

Token Lexer::lexIdentifier() {
    string value;
    while (isalnum(currentChar) && !isAtEnd()) {
        value += currentChar;
        advance();
    }
    return Token(TokenType::IDENTIFIER, value);
}

Token Lexer::lexNumber() {
    string value;
    while (isdigit(currentChar) && !isAtEnd()) {
        value += currentChar;
        advance();
    }
    return Token(TokenType::NUMBER, value);
}

Token Lexer::lexOperator() {
    char op = currentChar;
    advance();
    return Token(TokenType::OPERATOR, string(1, op));
}

File: JAVA/CompiladorLYC/src/compiladorlyc/codigo/Lexer.java
package compiladorlyc.codigo;

import java.util.ArrayList;
import java.util.List;

public class Lexer {
    private final String source;
    private int current;

    public Lexer(String source) {
        this.source = source;
        this.current = 0;
    }

    public List<Token> tokenize() {
        List<Token> tokens = new ArrayList<>();
        while (!isAtEnd()) {
            skipWhitespace();
            if (isAtEnd()) break; // Asegurarse de no avanzar más si se alcanza el final
            char c = advance();
            if (Character.isDigit(c)) {
                tokens.add(lexNumber());
            } else if (Character.isLetter(c)) {
                tokens.add(lexIdentifier());
            } else if ("+-*/=><".indexOf(c) >= 0) {
                tokens.add(lexOperator());
            } else if (c == ';') {
                tokens.add(lexSemicolon());
            } else {
                tokens.add(new Token(TokenType.UNKNOWN, String.valueOf(c)));
            }
        }
        tokens.add(new Token(TokenType.END_OF_FILE, ""));
        return tokens;
    }

    private boolean isAtEnd() {
        return current >= source.length();
    }

    private char advance() {
        if (!isAtEnd()) {
            return source.charAt(current++);
        }
        return '\0'; // Retorna un carácter nulo si ya está al final para evitar excepciones
    }

    private char peek() {
        if (isAtEnd()) return '\0';
        return source.charAt(current);
    }

    private Token lexNumber() {
        StringBuilder value = new StringBuilder();
        value.append(source.charAt(current - 1));
        while (Character.isDigit(peek()) || peek() == '.') {
            value.append(advance());
        }
        if (value.toString().contains(".")) {
            return new Token(TokenType.FLOAT, value.toString());
        } else {
            return new Token(TokenType.INTEGER, value.toString());
        }
    }

    private Token lexIdentifier() {
        StringBuilder value = new StringBuilder();
        value.append(source.charAt(current - 1));
        while (Character.isLetterOrDigit(peek())) {
            value.append(advance());
        }
        return new Token(TokenType.IDENTIFIER, value.toString());
    }

    private Token lexOperator() {
        StringBuilder value = new StringBuilder();
        value.append(source.charAt(current - 1));
        if ((value.toString().equals("=") || value.toString().equals(">") || value.toString().equals("<")) && peek() == '=') {
            value.append(advance());
        }
        return new Token(TokenType.OPERATOR, value.toString());
    }

    private Token lexSemicolon() {
        return new Token(TokenType.SEMICOLON, ";");
    }

    private void skipWhitespace() {
        while (!isAtEnd() && Character.isWhitespace(peek())) {
            advance();
        }
    }
}